### PR TITLE
Marcin

### PR DESCRIPTION
**Pull request by mjaelo**

### CODE REVIEW

File: src/editor/graph-editor.js [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+        mouseY = sigma.utils.getY(e);
+    }, false);
+    dom.addEventListener('DOMMouseScroll', function (e) {
+        radius *= sigma.utils.getDelta(e) < 0 ? 1 / wheelRatio : wheelRatio;
+    }, false);
+    dom.addEventListener('mousewheel', function (e) {
+        radius *= sigma.utils.getDelta(e) < 0 ? 1 / wheelRatio : wheelRatio;
+    }, false);
+    document.addEventListener('keydown', function (e) {
+        spaceMode = (e.which == 32) ? true : spaceMode;
```
Comment (Gunock) [inline src/editor/graph-editor.js:166]: Use === instead of ==.

File: src/editor/graph-editor.js [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+    });
+    document.addEventListener('keyup', function (e) {
+        spaceMode = e.which == 32 ? false : spaceMode;
```
Comment (Gunock) [inline src/editor/graph-editor.js:169]: Use === instead of ==.

Comment (mjaelo) [reply src/editor/graph-editor.js]: done

File: src/editor/graph-editor.js (hunk 1 above)
Comment (mjaelo) [reply src/editor/graph-editor.js]: done

File: src/editor/graph-editor.js [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+                ratio: 1 / scale
+            });
+
+            ground.style.top =
+                Math.max(h / 2 - Math.min((yMin + yMax) / 2 * scale, h), 0) + 'px';
+            disc.style.borderRadius = radius * scale;
+            disc.style.width = 2 * radius * scale;
+            disc.style.height = 2 * radius * scale;
+            disc.style.top = mouseY - radius * scale;
+            disc.style.left = mouseX - radius * scale;
```
[the remark this answers is not in the document]
Comment (mjaelo) [reply src/editor/graph-editor.js]: ur dum. lern Inglish

File: src/editor/graph-editor.js [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+    frame();
+
+
+    /**
+     * EVENTS BINDING:
+     * ***************
+     */
+    dom.addEventListener('click', function (e) {
+        // Find neighbors:
+        var x,
```
[the remark this answers is not in the document]
Comment (mjaelo) [reply src/editor/graph-editor.js]: done

File: src/editor/graph-editor.js [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+        s.graph.computePhysics();
+        s.refresh();
+
+        if (s.graph.nodes().length) {
+            var w = dom.offsetWidth,
+                h = dom.offsetHeight;
+
+            // The "rescale" middleware modifies the position of the nodes, but we
+            // need here the camera to deal with this. Here is the code:
+            var xMin = Infinity,
```
[the remark this answers is not in the document]
Comment (mjaelo) [reply src/editor/graph-editor.js]: done

File: src/editor/graph-editor.js [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+    disc = document.getElementById('disc');
+    ground = document.getElementById('ground');
+    c = s.camera;
+
+    function frame() {
+        s.graph.computePhysics();
+        s.refresh();
+
+        if (s.graph.nodes().length) {
+            var w = dom.offsetWidth,
```
[the remark this answers is not in the document]
Comment (mjaelo) [reply src/editor/graph-editor.js]: done

File: src/editor/graph-editor.js [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+    var s, c, dom, disc, ground, nId = 0, eId = 0, radius = 50, mouseX, mouseY, spaceMode = false, wheelRatio = 1.1,
+        nodeRadius = 10, inertia = 0.8, springForce = 0.01, springLength = 50, maxDisplacement = 15, gravity = 1.5;
+
+
+    /**
+     * CUSTOM PHYSICS LAYOUT:
+     * **********************
+     */
+    sigma.classes.graph.addMethod('computePhysics', function () {
+        var i, j, l = this.nodesArray.length, s, t, dX, dY, d, v;
```
[the remark this answers is not in the document]
Comment (mjaelo) [reply src/editor/graph-editor.js]: done

[the remark this answers is not in the document]
Comment (mjaelo) [reply src/editor/graph-editor.js]: done

File: src/editor/graph-editor.html [hunk 8]
```diff
@@ -0,0 +1,35 @@
+<html>
+<head></head>
```
[the remark this answers is not in the document]
Comment (mjaelo) [reply src/editor/graph-editor.html]: done

File: src/editor/graph-editor.html [hunk 9]
```diff
@@ -0,0 +1,35 @@
+<html>
```
[the remark this answers is not in the document]
Comment (mjaelo) [reply src/editor/graph-editor.html]: done

File: src/editor/graph-editor.html [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+<script src="renderers/webgl/sigma.webgl.nodes.def.js"></script>
+<script src="renderers/webgl/sigma.webgl.edges.def.js"></script>
+<script src="renderers/canvas/sigma.canvas.labels.def.js"></script>
+<script src="middlewares/sigma.middlewares.rescale.js"></script>
+<script src="middlewares/sigma.middlewares.copy.js"></script>
+<script src="misc/sigma.misc.bindEvents.js"></script>
+<script src="misc/sigma.misc.drawHovers.js"></script>
+<!-- END SIGMA IMPORTS -->
+<div id="container">
+    <link rel="stylesheet" type="text/css" href="graph-editor.css">
```
[the remark this answers is not in the document]
Comment (mjaelo) [reply src/editor/graph-editor.html]: done

File: src/editor/graph-editor.html [hunk 11]
```diff
@@ -0,0 +1,35 @@
+<html>
+<head></head>
+<body>
+<!-- START SIGMA IMPORTS -->
```
[the remark this answers is not in the document]
Comment (mjaelo) [reply src/editor/graph-editor.html]: done

File: src/editor/graph-editor.css [hunk 12]
```diff
@@ -0,0 +1,17 @@
+
+    #graph-container {
```
[the remark this answers is not in the document]
Comment (mjaelo) [reply src/editor/graph-editor.css]: done

File: src/editor/graph-editor.css [hunk 13]
```diff
@@ -0,0 +1,17 @@
+
```
[the remark this answers is not in the document]
Comment (mjaelo) [reply src/editor/graph-editor.css:1]: done

File: src/editor/graph-editor.js [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+
+            c.goTo({
+                x: (xMin + xMax) / 2,
+                y: (yMin + yMax) / 2,
+                ratio: 1 / scale
+            });
+
+            ground.style.top =
+                Math.max(h / 2 - Math.min((yMin + yMax) / 2 * scale, h), 0) + 'px';
+            disc.style.borderRadius = radius * scale;
```
[the remark this answers is not in the document]
Comment (mjaelo) [reply src/editor/graph-editor.js]: done

File: src/editor/graph-editor.js [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+            disc.style.width = 2 * radius * scale;
```
[the remark this answers is not in the document]
Comment (mjaelo) [reply src/editor/graph-editor.js]: done

File: src/editor/graph-editor.js [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+            disc.style.height = 2 * radius * scale;
```
[the remark this answers is not in the document]
Comment (mjaelo) [reply src/editor/graph-editor.js]: done

File: src/editor/graph-editor.js [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+            disc.style.top = mouseY - radius * scale;
```
[the remark this answers is not in the document]
Comment (mjaelo) [reply src/editor/graph-editor.js]: doe

**Pull request merged**